staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -214,33 +214,36 @@ func (wc *watchChan) startWatching(watchClosedCh chan struct{}) {
 
 	klog.V(3).Infof("Starting watcher for wc.ctx=%v, wc.key=%v", wc.ctx, wc.key)
 
-	updatedKey, opt := GetKeyAndOptFromPartitionConfig(wc.key, wc.partitionConfig)
-	wc.key = updatedKey
-	if opt != nil {
-		opts = append(opts, opt)
-	}
-
-	klog.V(3).Infof("Updated wc.key=%v and the opts is %+v", wc.key, opts)
-
-	wch := wc.watcher.client.Watch(wc.ctx, wc.key, opts...)
-	for wres := range wch {
-		if wres.Err() != nil {
-			err := wres.Err()
-			// If there is an error on server (e.g. compaction), the channel will return it before closed.
-			klog.Errorf("watch chan error: %v", err)
-			wc.sendError(err)
-			return
+	updatedKeys, updatedOptions := GetKeyAndOptFromPartitionConfig(wc.key, wc.partitionConfig)
+	for idx, updatedKey := range updatedKeys {
+		wc.key = updatedKey
+		opt := updatedOptions[idx]
+		updatedOpts := opts
+		if opt != nil {
+			updatedOpts = append(updatedOpts, opt)
 		}
-		for _, e := range wres.Events {
-			parsedEvent, err := parseEvent(e)
-			if err != nil {
+		wch := wc.watcher.client.Watch(wc.ctx, wc.key, opts...)
+		for wres := range wch {
+			if wres.Err() != nil {
+				err := wres.Err()
+				// If there is an error on server (e.g. compaction), the channel will return it before closed.
 				klog.Errorf("watch chan error: %v", err)
 				wc.sendError(err)
 				return
 			}
-			wc.sendEvent(parsedEvent)
+			for _, e := range wres.Events {
+				parsedEvent, err := parseEvent(e)
+				if err != nil {
+					klog.Errorf("watch chan error: %v", err)
+					wc.sendError(err)
```

Comment:
When there is an error, there might be retry. wc object could be reused. But key was already updated so it will have unexpected behavior. That's why I don't like update wc.key in line 219

Reply:
I removed it and keep it as the old way

Reply:
More investigation here.  All the keys go to the get operation which is stateless.

op := clientv3.OpGet(key, opts...)
	end := op.RangeBytes()
	pfxBegin, pfxEnd := prefixInterval(w.pfx, []byte(key), end)
	if pfxEnd != nil {
		opts = append(opts, clientv3.WithRange(string(pfxEnd)))
	}

	wch := w.Watcher.Watch(ctx, string(pfxBegin), opts...)